### PR TITLE
chore: Upgrade to Gateway API v1

### PR DESCRIPTION
**Pull request by Revolyssup**
### Type of change:

- [ ] Bugfix
- [ ] New feature provided
- [ ] Improve performance
- [ ] Backport patches
- [ ] Documentation
- [ ] Refactor
- [x] Chore
- [ ] CI/CD or Tests

### Pre-submission checklist:

- [x] Did you explain what problem does this PR solve? Or what new features have been added?
- [x] Have you added corresponding test cases?
- [ ] Have you modified the corresponding document?
- [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix-ingress-controller#community) first**

### CODE REVIEW

File: pkg/kube/gateway.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (Revolyssup) [inline pkg/kube/gateway.go:1]: if these structs are not used then delete this package

File: test/e2e/suite-chore/consistency.go [hunk 2]
```diff
@@ -180,8 +187,15 @@ var _ = ginkgo.Describe("suite-chore: Consistency between APISIX and the Ingress
 
 		s.NewAPISIXClient().GET("/ip").WithHeader("Host", "httpbin.org").Expect().Status(http.StatusOK)
 	})
-
+	if os.Getenv("K8S_VERSION") == "v1.29.0" {
```
Comment (AlinsRan) [inline test/e2e/suite-chore/consistency.go:190]: We may be able to obtain the k8 version through the API. please ref:
https://pkg.go.dev/k8s.io/client-go@v0.29.0/discovery#DiscoveryClient.ServerVersion


image


k8s version >= v1.25

**Comment by AlinsRan** [comment]
Oh..
There are too many workflow jobs, maybe we can integrate them in the future.

File: cmd/ingress/ingress.go [hunk 3]
```diff
@@ -177,6 +177,7 @@ For example, no available LB exists in the bare metal environment.`)
 	cmd.PersistentFlags().StringVar(&cfg.Kubernetes.IngressClass, "ingress-class", config.IngressClassApisixAndAll, "apisix-and-all is a special value, it handles Ingress resources with ingressClassName=apisix and all CRDs, the class of an Ingress object is set using the field IngressClassName in Kubernetes clusters version v1.18.0 or higher or the annotation \"kubernetes.io/ingress.class\" (deprecated)")
 	cmd.PersistentFlags().StringVar(&cfg.Kubernetes.ElectionID, "election-id", config.IngressAPISIXLeader, "election id used for campaign the controller leader")
 	cmd.PersistentFlags().StringVar(&cfg.Kubernetes.IngressVersion, "ingress-version", config.IngressNetworkingV1, "the supported ingress api group version, can be \"networking/v1beta1\", \"networking/v1\" (for Kubernetes version v1.19.0 or higher) and \"extensions/v1beta1\"")
+	cmd.PersistentFlags().StringVar(&cfg.Kubernetes.GatewayVersion, "gateway-version", config.GatewayNetworkingV1, "the supported gateway api group version, can be \"gateway/v1beta1\", \"gateway/v1\"")
```
[the remark this answers is not in the document]
Comment (Revolyssup) [reply cmd/ingress/ingress.go:180]: gateway-api-version sounds good to me

**Pull request closed without merging**